water/lib/Dialect/Wave/IR/WaveOps.cpp

```diff
@@ -110,6 +112,67 @@ llvm::LogicalResult wave::AllocateOp::verify() {
   return llvm::success();
 }
 
+//-----------------------------------------------------------------------------
+// ModuleOp
+//-----------------------------------------------------------------------------
+
+void wave::WaveModuleOp::build(mlir::OpBuilder &builder,
+                               mlir::OperationState &state,
+                               wave::WaveNormalForm normalForm,
+                               std::optional<llvm::StringRef> name) {
+  state.addRegion()->emplaceBlock();
+  state.addAttribute("normal_form",
+                     builder.getAttr<wave::WaveNormalFormAttr>(normalForm));
+  if (name) {
+    state.addAttribute(mlir::SymbolTable::getSymbolAttrName(),
+                       builder.getStringAttr(*name));
+  }
+}
+
+/// Construct a module from the given context.
+WaveModuleOp wave::WaveModuleOp::create(mlir::Location loc,
+                                        wave::WaveNormalForm normalForm) {
+  OpBuilder builder(loc->getContext());
+  return WaveModuleOp::create(builder, loc, normalForm);
+}
+
+/// Construct a module from the given context.
+WaveModuleOp wave::WaveModuleOp::create(mlir::Location loc,
+                                        wave::WaveNormalForm normalForm,
+                                        std::optional<llvm::StringRef> name) {
+  OpBuilder builder(loc->getContext());
+  return WaveModuleOp::create(builder, loc, normalForm, name);
+}
+
+llvm::LogicalResult wave::WaveModuleOp::verifyRegions() {
+  bool emitRemark = false;
+  bool emitDiagnostics = true;
+  if (!emitDiagnostics)
+    return llvm::success();
+
+  // assume or verify ?? child ops have been verified
```

Comment:
You can assume verifiers of child ops have run. This is the entire point of `verifyRegions`.